smaps/src/main/java/com/google/smaps/Histogram.java

```diff
@@ -0,0 +1,145 @@
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * Retrieves histogram information and formats it into a Json array for creating the chart
+ * and dashboard tools in histogram.js.
+ */
+@WebServlet(name = "Histogram", value = "/histogram")
+public class Histogram extends HttpServlet {
+  // Used to populate the lower bound text box and slider min.
+  static long lowerBound;
+  // Used to populate the upper bound text box and slider max.
+  static long upperBound;
+  // Flag indicating whether a user has entered info in textbox.
+  static boolean postFired = false;
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Get numbers from form in interactive-histogram.html.
+    String lower = request.getParameter("lower-bound");
+    String upper = request.getParameter("upper-bound");
+    postFired = true;
+
+    // In case user entered a number with .0 as the decimal, cast to long to drop the decimal,
+    // otherwise just convert the String to a Long.
+    if (lower.contains(".")) {
+      lowerBound = (long) Double.parseDouble(lower);
+    } else {
+      lowerBound = Long.parseLong(lower);
+    }
+    if (upper.contains(".")) {
+      upperBound = (long) Double.parseDouble(upper);
+    } else {
+      upperBound = Long.parseLong(upper);
+    }
+
+    // Reload the interactive-histogram.html.
+    response.sendRedirect("/interactive-histogram.html");
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Response will be a Json.
+    response.setContentType("application/json");
+
+    // TODO(sophbohr22): Change this from a hardcoded file to be the one the user uploads.
+    List<Region> regionList = FileParser.getRegionList("../smaps-full.txt");
+
+    // Parse histogram data.
+    List<Object[]> histogramData = makeDataArray(regionList);
+
+    // If the user hasn't entered custom bounds yet, use the min and max from regions list.
+    if (!postFired) {
+      setMinMax(regionList);
+    }
+
+    // Construct 2D array to hold the bounds.
+    long[] bounds = new long[2];
+    bounds[0] = lowerBound;
+    bounds[1] = upperBound;
+
+    // Transfer the Java Object arrays into JavaScript Objects (Json).
+    Gson boundsGson = new Gson();
+    Gson histGson = new Gson();
+    String histogramDataJson = histGson.toJson(histogramData);
+    String boundsJson = boundsGson.toJson(bounds);
+
+    // Put the two Json strings into a list to send to histogram.js.
+    List<String> jsonList = new ArrayList<String>();
+    jsonList.add(boundsJson);
+    jsonList.add(histogramDataJson);
+
+    // Write Json to histogram.js
+    response.getWriter().println(jsonList);
+  }
+
+  /** Creates list of 2D Object arrays of data for histogram. */
+  static ArrayList<Object[]> makeDataArray(List<Region> regions) {
```

Comment:
Why do these end up being of type Object? Is a more concrete type not known?

Reply:
To be able to send the Json array to Histogram.js, it has to contain the Strings for the axis titles in the same array as the values, so I went with Object.